main.js

```diff
@@ -138,18 +138,100 @@ if (args.reqJs)
         console.log(edge.toString());
     });
 
-if (args.cg) {
-    function pp(v) {
-        if (v.type === 'CalleeVertex')
-            return astutil.ppPos(v.call);
-        if (v.type === 'FuncVertex')
-            return astutil.ppPos(v.func);
-        if (v.type === 'NativeVertex')
-            return v.name;
-        throw new Error("strange vertex: " + v);
+if (args.cgPath) {
+    function constructNodeFromCallVertex(callVertex) {
+        if (!callVertex.call.attr.hasOwnProperty('enclosingFunction')) {
+            return {
+                'function_name': 'toplevel',
+                "file_name": callVertex.call.attr.enclosingFile,
+                'function_position': callVertex.call.loc.start.line + ":" + callVertex.call.loc.start.column,
+                'id':  callVertex.attr.node_id
+            }
+        }
+
+        return undefined;
+    }
+
+    function constructNodeFromFuncVertex(funcVertex) {
+        if (funcVertex.type === 'NativeVertex') {
+            return {
+                'function_name': funcVertex.name,
+                "file_name": 'native',
+                'function_position': '0:0',
+                'id':  funcVertex.attr.node_id
+            }
+        }
+
+        if (funcVertex.type === 'FuncVertex') {
+            return {
+                'function_name': funcVertex.func.attr.enclosingFile,
+                "file_name": funcVertex.func.id.name,
+                'function_position': funcVertex.func.id.loc.start.line + ":" + funcVertex.func.id.loc.start.column,
+                'id':  funcVertex.attr.node_id
+            }
+        }
+
+        return undefined;
     }
 
+    function addGraphNode(newNode, graphNodesById) {
+        if (newNode !== undefined && !(newNode.id in graphNodesById)) {
+            graphNodesById[newNode.id] = newNode;
+        }
+    }
+
+    function calleeName(callVertex) {
+        const callee = callVertex.call.callee;
+        if ('object' in callee && 'property' in callee) {
+            return callee.object.name + "." + callee.property.name;
+        }
+
+        return callee.name;
+    }
+
+    function addGraphEdge(callVertex, funcVertex, graphEdges) {
+        let sourceId = callVertex.attr.node_id;
+        if (callVertex.call.attr.hasOwnProperty('enclosingFunction')) {
+            // if the call is from a function, we use that function id as source id instead
+            sourceId = callVertex.call.attr.enclosingFunction.attr.func_vertex.attr.node_id;
```

Comment:
Are we sure that all properties will be always defined?  E.g. if `func_vertex` would be undefined, this code would throw an error. You can mitigate by using lodash and call `_.get()`.

Reply:
changed to use lodash